accepted/new-api.md

```diff
@@ -0,0 +1,971 @@
+
+  ```rust
+  trait HostFunctions {
+      async fn the_function(&self);
+  }
+  ```
+
+  The problem with this, though, is that `&self` is captured in the returned
+  future from `the_function`. For the returned future to be `Send` that then
+  requires `Self: Sync`, which then requires any mutable state in `Self` to have
+  locks or other forms of synchronization. The lowest-runtime-overhead way to
+  fix this is likely the usage of something similar to tokio's `TaskLocal`,
+  which means the trait would instead look like:
+
+  ```rust
+  trait HostFunctions {
+      async fn the_function(me: TaskLocal<Self>);
+  }
+  ```
+
+  but this makes accessing state in `me` much less ergonomic. With this proposal
+  the wiggle-generated trait would look like:
+
+  ```rust
+  trait HostFunctions<T> {
+      async fn the_function(store: StoreContextMut<'_, T>);
+  }
+  ```
+
+  where accessing state can simply be `store.get_mut()`.
+
+Note that these limitations above are only related to expressivity and where
+locks would be required (and some implementation requirements). This doesn't
+even start to get into how every single wasm operations would require taking and
+releasing a lock as well as accessing a thread-local to figure out what the
+current thread is. With wasmtime's per-function-call overhead at around ~20ns
+right now the addition of this overhead would likely at least double this
+overhead (and we would have no recourse to improve it).
+
+These limitations overall are what led to the conclusion that retrofitting
+`Send` and `Sync` on the current API is somewhat of a dead end. It's unclear if
+it can be feasibly implemented and the hoops that users of async would have to
+go through to get `Send` futures are quite numerous.
+
+# Open questions
+[open-questions]: #open-questions
+
+- It's not immediately clear if "pass in `&mut StoreContext<T>` to wasm get one
+  out the other end" is safe from Rust's type system perspective. Function calls
+  take `&mut StoreContext<T>` as an argument, and then provide the same `&mut
+  StoreContext<T>` as output. In the meantime, though, WebAssembly frames are on
+  the stack which have `*mut VMContext` pointers and such. It's believed that
+  this can be safely modeled in Rust's type system but it's not certain at this
+  time whether we'd need, for example, a few more `UnsafeCell`s internally or
+  whether we need to wrap something somewhere in an `UnsafeCell` or `*mut`
+  instead of `&mut`.
+
+- Is dropping support for the `wasm.h` C API feasible to do? Is this something
+  where we could influence the API itself upstream? Or is it reasonable for us
+  to continue implementing the existing C API?
+
+- If `ExternRef` moves to atomic reference counting and requires `T: Send +
+  Sync`, is that feasible? Are there use cases today that rely on where this
+  forces too much overhead? Along similar lines, are there possible designs for
+  `ExternRef` which relaxes some of these restrictions?
```

Comment:
Worth mentioning that Rust can borrow `ExternRef`s which means that even though we will be doing atomic ref counting, we won't be doing the ref count ops as much as you would otherwise in, say, C. And we can delay that incref until you actually need a new `ExternRef` handle and call `.clone()`. This makes me think that the switch to atomic ref counting won't be as devastating as it might otherwise have been.